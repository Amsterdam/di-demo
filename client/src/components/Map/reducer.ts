import { LatLngLiteral, Map } from 'leaflet';

interface IState {
    mapInstance?: Map | null;
    url?: string;
    query?: string;
    autosuggest?: ILocation[] | null;
    latLng?: LatLngLiteral | null;
    location?: ILocation | null;
    showAutosuggest?: boolean;
}

export interface ILocation {
    id?: string;
    displayName: string;
    latLng?: LatLngLiteral;
}

interface IAction {
    type: string;
    payload?: {
        mapInstance?: Map | null;
        url?: string;
        query?: string;
        autosuggest?: ILocation[] | null;
        location?: ILocation | null;
        showAutosuggest?: boolean;
    };
}

export const initialState: IState = {
    mapInstance: null,
    url: '',
    query: '',
    autosuggest: null,
    location: null,
    showAutosuggest: false
};

export const reducer = (state: IState = initialState, action: IAction): IState => {
    switch (action.type) {
        case 'setAutosuggest':
            return {
                ...state,
                autosuggest: (action.payload as any).autosuggest,
                showAutosuggest: true
            };

        case 'setLocation':
            return {
                ...state,
                location: (action.payload as any).location
            };

        case 'hideAutosuggest':
            return {
                ...state,
                showAutosuggest: false
            };

        case 'setMapInstance':
            return {
                ...state,
                mapInstance: (action.payload as any).mapInstance
            };

        case 'onChangeLocation':
            return {
                ...state,
                query: (action.payload as any).query,
                url: (action.payload as any).url
            };
        default:
            return state;
    }
};
