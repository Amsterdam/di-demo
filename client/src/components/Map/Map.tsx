import React, { useEffect, useCallback, useRef, useReducer } from 'react';
import 'leaflet/dist/leaflet.css';
import styled from 'styled-components';
import axios, { AxiosResponse } from 'axios';

import { initialState, reducer, Location } from './reducer';
import { Map, BaseLayer, ViewerContainer, Zoom, Marker } from '@datapunt/arm-core';
import { Input } from '@datapunt/asc-ui';
import { Link, ListItem, Icon, themeColor, themeSpacing } from '@datapunt/asc-ui';
import { ChevronRight } from '@datapunt/asc-assets';
import { LatLng, LeafletMouseEvent } from 'leaflet';

interface IProps {
    updateLocationCallback: (location: Location[]) => void;
}

const MapComponent: React.FC<IProps> = ({ updateLocationCallback }) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    const locationRef = useRef(null);
    const wrapperRef = useRef(null);
    const mapRef = useRef(null);
    const autosuggestUrl =
        'https://geodata.nationaalgeoregister.nl/locatieserver/v3/suggest?fq=gemeentenaam:amsterdam&fq=type:adres&fl=id,weergavenaam,type,score,lat,lon&q=';
    const locationUrl =
        'https://geodata.nationaalgeoregister.nl/locatieserver/revgeo?type=adres&rows=1&fl=id,weergavenaam,straatnaam,huis_nlt,postcode,woonplaatsnaam,centroide_ll&distance=50&';
    const lookupUrl = 'https://geodata.nationaalgeoregister.nl/locatieserver/v3/lookup?id=';

    const { mapInstance, url, query, autosuggest, latLng, location, showAutosuggest } = state;

    const fetchAutosuggest = useCallback(
        async url => {
            const response: AxiosResponse = await axios.get(url);
            dispatch({
                type: 'setAutosuggest',
                payload: {
                    autosuggest: response.data?.response?.docs
                }
            });
        },
        [dispatch]
    );

    const fetchLocation = useCallback(
        async (loc: LatLng) => {
            const response = await axios.get(`${locationUrl}&lat=${loc.lat}&lon=${loc.lng}`);
            const location = response.data?.response?.docs;
            dispatch({
                type: 'setLocation',
                payload: {
                    location
                }
            });

            if (updateLocationCallback) {
                updateLocationCallback(location.length === 0 ? [{ weergavenaam: 'geenAdresGevonden' }] : location);
            }
        },
        [locationUrl, dispatch, updateLocationCallback]
    );

    const onMapClick = useCallback(
        async (e: LeafletMouseEvent) => {
            dispatch({
                type: 'setLatLng',
                payload: {
                    latLng: e.latlng
                }
            });

            fetchLocation(e.latlng);
        },
        [dispatch, fetchLocation]
    );

    const onAutosuggestClick = useCallback(
        async (e: React.SyntheticEvent<LeafletMouseEvent>, location: Location) => {
            e.preventDefault();
            if (location.weergavenaam) {
                locationRef.current.value = location.weergavenaam;
                dispatch({
                    type: 'hideAutosuggest'
                });
            }

            const response = await axios.get(`${lookupUrl}${location.id}`);
            if (mapInstance && response.data.response.docs[0]) {
                const loc = response.data.response.docs[0].centroide_ll.replace(/POINT\(|\)/, '').split(' ');
                const targetLatlng = { lat: parseFloat(loc[1]), lng: parseFloat(loc[0]) };
                mapInstance.flyTo(targetLatlng, 11);
                dispatch({
                    type: 'setLatLng',
                    payload: {
                        latLng: targetLatlng
                    }
                });
                if (updateLocationCallback) {
                    updateLocationCallback(response.data.response.docs);
                }
            }
        },
        [mapInstance, lookupUrl, dispatch, updateLocationCallback]
    );

    useEffect(() => {
        if (url.length) {
            fetchAutosuggest(url);
        }
    }, [url, fetchAutosuggest]);

    useEffect(() => {
        locationRef.current.value = location && location.length ? location[0].weergavenaam : '';
    }, [location]);

    const handleClickOutside = useCallback(
        e => {
            if (wrapperRef.current && !wrapperRef.current.contains(e.target)) {
                dispatch({
                    type: 'hideAutosuggest'
                });

                locationRef.current.value = '';
            } else if (locationRef.current && locationRef.current.contains(e.target)) {
                // do nothing
            } else if (mapRef.current && mapRef.current.contains(e.target)) {
                const latlng = mapInstance.mouseEventToLatLng(e);
                onMapClick({ ...e, latlng });
            }
        },
        [mapInstance, dispatch, onMapClick]
    );

    useEffect(() => {
        if (mapInstance) {
            document.body.addEventListener('click', handleClickOutside, false);
        }

        return () => {
            document.body.removeEventListener('click', handleClickOutside, false);
        };
    }, [mapInstance, handleClickOutside]);

    return (
        <MapParent ref={mapRef}>
            <StyledMap
                data-testid="map"
                setInstance={instance => {
                    dispatch({
                        type: 'setMapInstance',
                        payload: {
                            mapInstance: instance
                        }
                    });
                }}
                events={{
                    click: (e: LeafletMouseEvent) => {
                        onMapClick(e);
                    }
                }}
            >
                {latLng && <Marker latLng={latLng} />}
                <StyledViewerContainer
                    bottomRight={<Zoom />}
                    topLeft={
                        <>
                            <StyledInput
                                id="location"
                                data-testid="input"
                                ref={locationRef}
                                onChange={e => {
                                    if (e.target.value.length < 3) return;
                                    const value = encodeURIComponent(e.target.value);
                                    dispatch({
                                        type: 'onChangeLocation',
                                        payload: {
                                            query: e.target.value,
                                            url: `${autosuggestUrl}${value}`
                                        }
                                    });
                                }}
                            />
                            {showAutosuggest && query.length && autosuggest && autosuggest.length ? (
                                <StyledAutosuggest data-testid="autosuggest" ref={wrapperRef}>
                                    {autosuggest.map(item => (
                                        <ListItem key={item.id}>
                                            <StyledIcon size={14}>
                                                <ChevronRight />
                                            </StyledIcon>
                                            <Link href="#" variant="inline" onClick={e => onAutosuggestClick(e, item)}>
                                                {item.weergavenaam}
                                            </Link>
                                        </ListItem>
                                    ))}
                                </StyledAutosuggest>
                            ) : null}
                        </>
                    }
                />
                <BaseLayer />
            </StyledMap>
        </MapParent>
    );
};

const MapParent = styled.div`
    position: relative;
    height: 500px;
    width: 100%;
    margin-bottom: ${themeSpacing(5)};
`;

const StyledMap = styled(Map)`
    height: 500px;
    width: 100%;
    cursor: pointer;
    z-index: 0;
`;

const StyledViewerContainer = styled(ViewerContainer)`
    > div {
        width: calc(100% - ${themeSpacing(8)});
    }
    z-index: 1;
`;

const StyledInput = styled(Input)`
    width: 100%;
    margin-bottom: -17px;
`;

const StyledAutosuggest = styled.ul`
    width: 100%;
    background-color: ${themeColor('tint', 'level1')};
    list-style-type: none;
    padding: 6px 0 0 ${themeSpacing(3)};
    border: 1px solid ${themeColor('tint', 'level5')};
    a {
        display: inline;
        color: ${themeColor('tint', 'level7')};
    }
`;

const StyledIcon = styled(Icon)`
    display: inline;
    margin-right: 8px;
`;

export default MapComponent;
