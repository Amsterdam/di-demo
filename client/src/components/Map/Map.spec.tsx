import React from 'react';
import { screen, act, fireEvent } from '@testing-library/react';
import { setupMocks, wrappedRender } from '@test/utils';
import Map from '@components/Map/Map';

// Setup all the generic mocks
setupMocks();

// MockcreateIrmaSession
jest.mock('@services/createIrmaSession');

describe('Map', () => {
    // it.skip('should render default map with input and no autosuggest visible', async () => {
    //     await act(async () => await wrappedRender(<Map updateLocationCallback={jest.fn()} />));

    //     expect(screen.queryByTestId('map')).toBeInTheDocument();
    //     expect(screen.queryByTestId('input')).toBeInTheDocument();
    //     expect(screen.queryByTestId('autosuggest')).not.toBeInTheDocument();
    // });

    it('should show autosuggest when input field has changed', async () => {
        const spy = jest.fn();
        await act(async () => await wrappedRender(<Map updateLocationCallback={jest.fn} />));

        act(() => {
            fireEvent.change(screen.queryByTestId('input'), { target: { value: 'Dam 1' } });
        });

        act(() => {
            expect(screen.getByTestId('autosuggest')).toBeInTheDocument();
        });
    });

    // it.skip('should show correct address when clicked on th map', async () => {
    //     const spy = jest.fn();

    //     await act(async () => await wrappedRender(<Map updateLocationCallback={jest.fn} />));

    //     await act(
    //         async () =>
    //             await fireEvent.click(screen.queryByTestId('map'), {
    //                 aaa: '123'
    //             })
    //     );

    //     act(() => {
    //         // expect(screen.getByTestId('map')).toMatchSnapshot();
    //         expect(spy).toHaveBeenCalled();
    //     });
    // });
});
