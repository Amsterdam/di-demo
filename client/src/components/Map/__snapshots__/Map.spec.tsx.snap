// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map should render default map with input and no autosuggest visible 1`] = `
.c0 {
  height: 500px;
  width: 100%;
  cursor: pointer;
}

<div
  class="c0 leaflet-container leaflet-grab leaflet-touch-drag"
  data-testid="map"
  style="position: relative;"
  tabindex="0"
>
  <div
    class="leaflet-pane leaflet-map-pane"
    style="left: 0px; top: 0px;"
  >
    <div
      class="leaflet-pane leaflet-tile-pane"
    >
      <div
        class="leaflet-layer "
        style="z-index: 1;"
      >
        <div
          class="leaflet-tile-container leaflet-zoom-animated"
          style="z-index: 18; left: 0px; top: 0px;"
        >
          <img
            alt=""
            class="leaflet-tile"
            role="presentation"
            src="https://t4.data.amsterdam.nl/topo_rd/10/472/540.png"
            style="width: 256px; height: 256px; left: -230px; top: -173px;"
          />
        </div>
      </div>
    </div>
    <div
      class="leaflet-pane leaflet-shadow-pane"
    />
    <div
      class="leaflet-pane leaflet-overlay-pane"
    />
    <div
      class="leaflet-pane leaflet-marker-pane"
    />
    <div
      class="leaflet-pane leaflet-tooltip-pane"
    />
    <div
      class="leaflet-pane leaflet-popup-pane"
    />
  </div>
  <div
    class="leaflet-control-container"
  >
    <div
      class="leaflet-top leaflet-left"
    />
    <div
      class="leaflet-top leaflet-right"
    />
    <div
      class="leaflet-bottom leaflet-left"
    />
    <div
      class="leaflet-bottom leaflet-right"
    >
      <div
        class="leaflet-control-attribution leaflet-control"
      >
        <a
          href="https://leafletjs.com"
          title="A JS library for interactive maps"
        >
          Leaflet
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Map should show autosuggest when input field has changed 1`] = `
.c3 {
  margin: 0;
  color: #000000;
  font-weight: inherit;
  font-size: 16px;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  line-height: 22px;
  margin-bottom: 0;
  font-stretch: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c4 {
  display: inline-block;
  color: #004699;
}

.c4:focus {
  background-color: #FEC813;
}

.c4:hover {
  color: #ec0000;
}

.c1 {
  word-break: break-word;
  margin-bottom: 8px;
}

.c0 {
  width: 500px;
  background-color: #ffffff;
  list-style-type: none;
  padding: 6px 0 0 12px;
  border: 1px solid #767676;
}

.c0 a {
  color: #000000;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 14px;
  height: 14px;
  box-sizing: content-box;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  display: inline;
  margin-right: 8px;
}

.c2 > svg {
  width: 14px;
  height: 14px;
}

@media screen and (min-width:1024px) {
  .c3 {
    font-size: 18px;
    line-height: 25px;
  }
}

<ul
  class="c0"
  data-testid="autosuggest"
>
  <li
    class="c1"
  >
    <span
      class="c2"
      size="14"
    >
      <svg
        focusable="false"
        viewBox="0 0 32 32"
      >
        <path
          d="M9.757 32l-2.9-2.91L19.937 16 6.857 2.91 9.757 0l16 16z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <a
      class="c3 c4"
      href="#"
    >
      Dam 1, 1012JS Amsterdam
    </a>
  </li>
  <li
    class="c1"
  >
    <span
      class="c2"
      size="14"
    >
      <svg
        focusable="false"
        viewBox="0 0 32 32"
      >
        <path
          d="M9.757 32l-2.9-2.91L19.937 16 6.857 2.91 9.757 0l16 16z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <a
      class="c3 c4"
      href="#"
    >
      Damrak 1, 1012LG Amsterdam
    </a>
  </li>
</ul>
`;
